# <font color=blue>Assignments for "Evaluating Goodness of Fit"</font>

As in previous lessons, please submit a link to a single gist that contains links to two Juypyter notebooks (one for each assignment below).

## 1. Weather model

For this assignment, you'll revisit the historical temperature dataset. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import Normalizer

%matplotlib inline

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

- Like in the previous lesson, build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why? 

In [3]:
y = weather_df['Temperature (C)'] - weather_df['Apparent Temperature (C)']

X = weather_df[['Humidity', 'Wind Speed (km/h)']]

In [4]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Wed, 12 May 2021   Prob (F-statistic):               0.00
Time:                        22:15:58   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Both R-squared and adjusted R-squared are 0.288, which are not satisfactory. Usually, an R-squared value of 0.7 and above are satisfactory. An R-squared value of 0.3 and below are indeed not good at all.

- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one? 


In [5]:
X['humidity*windspeed'] = X['Humidity'] * X['Wind Speed (km/h)']

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Wed, 12 May 2021   Prob (F-statistic):               0.00
Time:                        22:15:58   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity              -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)     -0.0905      0.002    -36.797      0.000      -0.095      -0.086
humidity*windspeed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Now the R-squared value is 0.341, which is an improvement to the previous model and it is now at least higher than 0.3, which is can be thought of as a minimum threshold for R-squared values.

- Add *visibility* as additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the *visibility* in terms of the improvement in the adjusted R-squared. Which one is more useful?

In [6]:
X['visibility'] = weather_df['Visibility (km)']
X.drop('humidity*windspeed',axis=1, inplace=True)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Wed, 12 May 2021   Prob (F-statistic):               0.00
Time:                        22:15:58   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.5756      0.028    -56.605      0.000      -1.630      -1.521
Humidity              2.6066      0.025    102.784      0.000       2.557       2.656
Wind Speed (km/h)     0.1199      0.001    179.014      0.000       0.119       0.121
visibility           -0.0540      0.001    -46.614      0.000      -0.056      -0.052
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                           0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Although both R-squared and adjusted R-squared have increased by comparison with the first model, the interaction term resulted in a better (adjusted) R-squared in the previous model.

- Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

For both AIC and BIC, the lower value the better. Therefore, the second model, that is, the model with the interaction term was better with respect to their AIC and BIC scores.

##  2. House prices model

In this exercise, you'll work on your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [7]:
df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

- Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.

In [8]:
df_null_ratios = df.isna().sum() / len(df)
df.drop(df_null_ratios[df_null_ratios > .1].index, axis=1, inplace=True)
df.dropna(inplace=True)

X_num = df[['OverallQual', 'GrLivArea', 'GarageCars', 'TotalBsmtSF']]

ExterQual = pd.get_dummies(df['ExterQual'], drop_first=True)
BsmtQual = pd.get_dummies(df['BsmtQual'], drop_first=True)
BsmtFinType1 = pd.get_dummies(df['BsmtFinType1'], drop_first=True)
KitchenQual = pd.get_dummies(df['KitchenQual'], drop_first=True)
GarageQual = pd.get_dummies(df['GarageQual'], drop_first=True)
GarageCond = pd.get_dummies(df['GarageCond'], drop_first=True)

X_cat = pd.concat([ExterQual, BsmtQual, BsmtFinType1, KitchenQual, GarageQual, GarageCond], axis=1)

scaler = StandardScaler()

X_num = pd.DataFrame(scaler.fit_transform(X_num), columns=X_num.columns, index=X_num.index)

X = pd.concat([X_num, X_cat], axis=1)
y = df['SalePrice']

X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.804
Model:                            OLS   Adj. R-squared:                  0.800
Method:                 Least Squares   F-statistic:                     207.1
Date:                Wed, 12 May 2021   Prob (F-statistic):               0.00
Time:                        22:15:59   Log-Likelihood:                -15895.
No. Observations:                1338   AIC:                         3.184e+04
Df Residuals:                    1311   BIC:                         3.198e+04
Df Model:                          26                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const        2.936e+05   2.59e+04     11.323      0.000    2.43e+05    3.44e+05
OverallQual  1.931e+04   1782.349     10.831      0.000    1.58e+04    2.28e+04
GrLivArea    2.578e+04   1281.902     20.109      0.000    2.33e+04    2.83e+04
GarageCars   9857.3916   1323.699      7.447      0.000    7260.592    1.25e+04
TotalBsmtSF  7218.1581   1250.482      5.772      0.000    4764.994    9671.322
Fa          -4.274e+04   1.55e+04     -2.762      0.006   -7.31e+04   -1.24e+04
Gd          -1.759e+04   6579.440     -2.674      0.008   -3.05e+04   -4684.228
TA          -2.523e+04   7281.624     -3.466      0.001   -3.95e+04    -1.1e+04
Fa          -4.172e+04   8373.926     -4.982      0.000   -5.81e+04   -2.53e+04
Gd          -3.462e+04   4458.877     -7.764      0.000   -4.34e+04   -2.59e+04
TA          -4.012e+04   5211.001     -7.698      0.000   -5.03e+04   -2.99e+04
BLQ            47.7856   3891.035      0.012      0.990   -7585.549    7681.120
GLQ          2502.4555   3395.036      0.737      0.461   -4157.842    9162.753
LwQ          -1.15e+04   4946.616     -2.324      0.020   -2.12e+04   -1793.668
Rec         -5396.4850   4061.925     -1.329      0.184   -1.34e+04    2572.099
Unf         -1.765e+04   3219.891     -5.482      0.000    -2.4e+04   -1.13e+04
Fa          -3.717e+04   9312.861     -3.992      0.000   -5.54e+04   -1.89e+04
Gd          -2.789e+04   4962.890     -5.620      0.000   -3.76e+04   -1.82e+04
TA          -3.676e+04   5545.114     -6.629      0.000   -4.76e+04   -2.59e+04
Fa          -1.019e+05   3.65e+04     -2.790      0.005   -1.74e+05   -3.03e+04
Gd          -8.489e+04   3.73e+04     -2.274      0.023   -1.58e+05   -1.17e+04
Po           -9.37e+04   4.57e+04     -2.048      0.041   -1.83e+05   -3953.421
TA          -9.628e+04    3.6e+04     -2.673      0.008   -1.67e+05   -2.56e+04
Fa            7.32e+04   4.45e+04      1.644      0.100   -1.41e+04    1.61e+05
Gd           8.044e+04   4.56e+04      1.762      0.078   -9104.434     1.7e+05
Po           6.444e+04   4.79e+04      1.344      0.179   -2.96e+04    1.59e+05
TA           8.136e+04   4.39e+04      1.852      0.064   -4812.496    1.68e+05
==============================================================================
Omnibus:                      708.209   Durbin-Watson:                   1.974
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            85127.494
Skew:                          -1.460   Prob(JB):                         0.00
Kurtosis:                      41.967   Cond. No.                         240.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- F-test: 207.1
- R-squared: 0.804
- Adjusted R-squared: 0.800
- AIC: 31840.0
- BIC: 31980.0

- Do you think your model is satisfactory? If so, why?

Looking at all the criteria discussed so far, especially (adjusted) R-squared and F-test, I think my model is satisfactory.

- In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables. 

In [9]:
# Including all numerical features, not just the ones with highest
# correlations, which was the case in the previous model.

df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")

df_null_ratios = df.isna().sum() / len(df)
df.drop(df_null_ratios[df_null_ratios > .1].index, axis=1, inplace=True)
df.dropna(inplace=True)

X_num = df.select_dtypes(exclude='O').drop('SalePrice', axis=1)

ExterQual = pd.get_dummies(df['ExterQual'], drop_first=True)
BsmtQual = pd.get_dummies(df['BsmtQual'], drop_first=True)
BsmtFinType1 = pd.get_dummies(df['BsmtFinType1'], drop_first=True)
KitchenQual = pd.get_dummies(df['KitchenQual'], drop_first=True)
GarageQual = pd.get_dummies(df['GarageQual'], drop_first=True)
GarageCond = pd.get_dummies(df['GarageCond'], drop_first=True)

X_cat = pd.concat([ExterQual, BsmtQual, BsmtFinType1, KitchenQual, GarageQual, GarageCond], axis=1)

scaler = StandardScaler()

X_num = pd.DataFrame(scaler.fit_transform(X_num), columns=X_num.columns, index=X_num.index)

X = pd.concat([X_num, X_cat], axis=1)
y = df['SalePrice']

X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.838
Model:                            OLS   Adj. R-squared:                  0.831
Method:                 Least Squares   F-statistic:                     118.0
Date:                Wed, 12 May 2021   Prob (F-statistic):               0.00
Time:                        22:16:01   Log-Likelihood:                -15769.
No. Observations:                1338   AIC:                         3.165e+04
Df Residuals:                    1281   BIC:                         3.195e+04
Df Model:                          56                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          2.859e+05   2.44e+04     11.709      0.000    2.38e+05    3.34e+05
Id              -83.6026    908.402     -0.092      0.927   -1865.721    1698.516
MSSubClass    -7050.3573   1104.362     -6.384      0.000   -9216.915   -4883.800
LotArea        5058.9383    992.586      5.097      0.000    3111.665    7006.211
OverallQual    1.546e+04   1749.607      8.834      0.000     1.2e+04    1.89e+04
OverallCond    6435.5934   1199.783      5.364      0.000    4081.839    8789.348
YearBuilt      9658.3546   2357.791      4.096      0.000    5032.799    1.43e+04
YearRemodAdd    480.6198   1543.878      0.311      0.756   -2548.188    3509.427
MasVnrArea     3546.3958   1090.179      3.253      0.001    1407.664    5685.127
BsmtFinSF1      784.0917   1171.019      0.670      0.503   -1513.233    3081.417
BsmtFinSF2      514.0832    994.712      0.517      0.605   -1437.361    2465.527
BsmtUnfSF       665.9217   1123.767      0.593      0.554   -1538.704    2870.547
TotalBsmtSF    1820.0930   1684.460      1.081      0.280   -1484.511    5124.697
1stFlrSF       7616.0928   1763.737      4.318      0.000    4155.963    1.11e+04
2ndFlrSF       9654.4123   1336.466      7.224      0.000    7032.511    1.23e+04
LowQualFinSF  -1002.6219   1011.294     -0.991      0.322   -2986.597     981.353
GrLivArea      1.374e+04   1413.580      9.723      0.000     1.1e+04    1.65e+04
BsmtFullBath   3511.3976   1391.321      2.524      0.012     781.879    6240.916
BsmtHalfBath     -0.4557    967.484     -0.000      1.000   -1898.482    1897.571
FullBath       2629.9822   1604.961      1.639      0.102    -518.658    5778.622
HalfBath       -329.6007   1355.421     -0.243      0.808   -2988.690    2329.488
BedroomAbvGr  -3762.5746   1409.566     -2.669      0.008   -6527.886    -997.263
KitchenAbvGr  -3723.0147   1101.791     -3.379      0.001   -5884.527   -1561.503
TotRmsAbvGrd   4111.7755   1972.105      2.085      0.037     242.865    7980.686
Fireplaces     2968.8136   1127.259      2.634      0.009     757.337    5180.290
GarageYrBlt   -2873.0402   1973.228     -1.456      0.146   -6744.154     998.074
GarageCars     9475.1257   1857.839      5.100      0.000    5830.384    1.31e+04
GarageArea     -168.9439   1845.939     -0.092      0.927   -3790.339    3452.451
WoodDeckSF     3053.1682    988.617      3.088      0.002    1113.682    4992.655
OpenPorchSF   -1841.8884   1001.500     -1.839      0.066   -3806.649     122.872
EnclosedPorch   755.1437   1015.870      0.743      0.457   -1237.808    2748.096
3SsnPorch       520.1041    906.768      0.574      0.566   -1258.810    2299.018
ScreenPorch    2615.7745    956.945      2.733      0.006     738.423    4493.126
PoolArea      -1228.1424    953.899     -1.287      0.198   -3099.519     643.234
MiscVal        -222.6602    905.428     -0.246   

- For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

Initial model:
- R-squared:	0.804
- Adj. R-squared:	0.800
- F-statistic:	207.1
- Prob (F-statistic):	0.00
- AIC:	31840.0
- BIC:	31980.0

Improved model:
- R-squared:	0.838
- Adj. R-squared:	0.831
- F-statistic:	118.0
- Prob (F-statistic):	0.00
- AIC:	31650.0
- BIC:	31950.0

In the first model, I have included only 4 numerical feature with the highest correlations. However, in the second model, I have included all the numerical features, which resulted in a better model.